src/main/developerMode.ts

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {ipcMain} from 'electron';
+import {EventEmitter} from 'events';
+
+import {DEVELOPER_MODE_UPDATED, IS_DEVELOPER_MODE_ENABLED, UPDATE_PATHS, GET_DEVELOPER_MODE_SETTING} from 'common/communication';
+import JsonFileManager from 'common/JsonFileManager';
+import {developerModeJson} from 'main/constants';
+
+import type {DeveloperSettings} from 'types/settings';
+
+export class DeveloperMode extends EventEmitter {
+    private json: JsonFileManager<DeveloperSettings>;
+
+    constructor(file: string) {
+        super();
+        this.json = new JsonFileManager(file);
+
+        ipcMain.handle(IS_DEVELOPER_MODE_ENABLED, this.enabled);
+        ipcMain.handle(GET_DEVELOPER_MODE_SETTING, (_, setting) => this.get(setting));
+    }
+
+    enabled = () => process.env.MM_DESKTOP_DEVELOPER_MODE === 'true';
+
+    toggle = (setting: keyof DeveloperSettings) => {
+        if (!this.enabled()) {
+            return;
+        }
+
+        this.json.setValue(setting, !this.json.getValue(setting));
+        this.emit(DEVELOPER_MODE_UPDATED, {[setting]: this.json.getValue(setting)});
+    };
+
+    get = (setting: keyof DeveloperSettings) => {
+        if (!this.enabled()) {
+            return false;
+        }
+
+        return this.json.getValue(setting);
+    };
+
+    switchOff = (
+        setting: keyof DeveloperSettings,
+        onStart: () => void,
+        onStop: () => void,
+    ) => {
+        if (!this.get(setting)) {
+            onStart();
+        }
```

Comment:
(0/5 and kind of ranty) I wonder if we might want to also call `onStop` when this method is called with the feature disabled to help reduce the number of states that the app is in. Currently, features that use this can be in one of three states:

1. `onStart` was called last because `switchOff` was called with the setting enabled
2. `onStop` was called last because `switchOff` was called with the setting enabled and then the setting was disabled
3. Neither `onStop` not `onStart` have been called because `switchOff` was called with the setting disabled

The third state is potentially the weird one if `onStop` puts the feature into a different state than its initial state. While it might seem a bit odd to stop something that was never started, always calling `onStop` gets rid of that rare third case so we never have to think about it